### PR TITLE
WordData forms reinstated...

### PR DESCRIPTION
**Pull request by rocky**
Make better use if Symbols and some correct type mismatches

Probably the code be revised so that to_expression, and to_mathics_list are not needed. That is deferred for another PR.

### CODE REVIEW

File: pymathics/natlang/main.py [hunk 1]
```diff
@@ -367,8 +388,10 @@ class DeleteStopwords(_SpacyBuiltin):
       <dd>returns $string$ without stopwords.
     </dl>
 
-    >> DeleteStopwords[{"Somewhere", "over", "the", "rainbow"}]
-     = {rainbow}
+    ## This has changed since old versions of natlang, and I am
+    ## I am not sure the old behavior was correct.
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:392]: There are 2 "I am"

File: pymathics/natlang/main.py [hunk 2]
```diff
@@ -349,7 +370,7 @@ def eval(self, text: String, evaluation: Evaluation, options: dict):
             return ListExpression(*[String(sent.text) for sent in doc.sents])
 
     def eval_n(self, text: String, n: Integer, evaluation: Evaluation, options: dict):
-        "TextSentences[text_String, n_Integer, OptionsPattern[%(name)s]]"
+        "TextSentences[text_String, n_Integer, OptionsPattern[TextSentences]]"
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:373]: Why this change?

File: pymathics/natlang/main.py [hunk 3]
```diff
@@ -1139,36 +1158,36 @@ def _standard_property(
...
-            return Expression(SymbolMissing, "NotAvailable")
+            return Expression(SymbolMissing, StringNotAvailable)
         elif py_form == "List":
             word_property = WordProperty(self._short_syn_form, wordnet, language_code)
             property_getter = getattr(
                 word_property, "%s" % self._underscore(py_property), None
             )
             if property_getter:
-                return ListExpression(
+                return to_mathics_list(
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:1168]: Is `to_mathics_list` preferred now?

File: pymathics/natlang/main.py [hunk 4]
```diff
@@ -1203,8 +1222,8 @@ def _property(
         except MessageException as e:
             e.message(evaluation)
 
-    def eval(self, word, evaluation: Evaluation, options: dict):
-        "WordData[word_, OptionsPattern[%(name)s]]"
+    def eval(self, word, evaluation: Evaluation, options: dict) -> Optional[Expression]:
+        "WordData[word_, OptionsPattern[WordData]]"
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:1226]: Why remove `%(name)s`?

File: pymathics/natlang/main.py [hunk 5]
```diff
@@ -701,10 +724,7 @@ def eval_pair(self, text1, i1, text2, i2, evaluation: Evaluation, options: dict)
             else:
                 doc2 = self._nlp(text2.value, evaluation, options)
             if doc2:
-                if (
-                    i1.get_head_name() == "System`List"
-                    and i2.get_head_name() == "System`List"
-                ):
+                if i1.get_head() == SymbolList and i2.get_head() == SymbolList:
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:727]: Can't `is` be used instead of `==` here?

File: pymathics/natlang/main.py [hunk 6]
```diff
@@ -1203,8 +1222,8 @@ def _property(
         except MessageException as e:
             e.message(evaluation)
 
-    def eval(self, word, evaluation: Evaluation, options: dict):
-        "WordData[word_, OptionsPattern[%(name)s]]"
+    def eval(self, word, evaluation: Evaluation, options: dict) -> Optional[Expression]:
+        "WordData[word_, OptionsPattern[WordData]]"
         if word.get_head_name() == "System`StringExpression":
             return Expression(SymbolDictionaryLookup, word)
         elif isinstance(word, String) or word.get_head_name() == "System`List":
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:1229]: Maybe `get_head` here and above?

File: pymathics/natlang/main.py [hunk 7]
```diff
@@ -1223,23 +1242,24 @@ def eval(self, word, evaluation: Evaluation, options: dict):
...
 
     def eval_property(self, word, property, evaluation: Evaluation, options: dict):
-        "WordData[word_, property_String, OptionsPattern[%(name)s]]"
-        if word.get_head_name() == "System`StringExpression":
+        "WordData[word_, property_String, OptionsPattern[WordData]]"
+        if word.get_head is SymbolStringExpression:
             if property.get_string_value() == "Lookup":
                 return Expression(SymbolDictionaryLookup, word)
-        elif isinstance(word, String) or word.get_head_name() == "System`List":
+        elif isinstance(word, String) or word.get_head() is SymbolList:
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:1253]: It seems like there is an inconsistency in `get_head sometimes it is an attribute and sometimes it is a method.
BTW, in the attribute case I think it should be `head` instead of `get_head`

File: pymathics/natlang/main.py [hunk 8]
```diff
@@ -1223,23 +1242,24 @@ def eval(self, word, evaluation: Evaluation, options: dict):
...
             return self._property(
                 word, property.get_string_value(), "ShortRules", evaluation, options
             )
 
     def eval_property_form(
         self, word, property, form, evaluation: Evaluation, options: dict
     ):
-        "WordData[word_, property_String, form_String, OptionsPattern[%(name)s]]"
-        if isinstance(word, String) or word.get_head_name() == "System`List":
+        "WordData[word_, property_String, form_String, OptionsPattern[WordData]]"
```
Comment (TiagoCavalcante) [inline pymathics/natlang/main.py:1261]: Why remove `%(name)s`?

**Review by TiagoCavalcante — APPROVED** [review]
LGTM

File: pymathics/natlang/main.py (hunk 2 above)
Comment (rocky) [reply pymathics/natlang/main.py:373]: The short answer is clarity over cleverness.

There are levels of rules or conventions going on here.

The first convention is Mathics3 Case pattern matching which is used in rule application to figure out which class method to call. Stuff like `_String` fall into that category. 

And the second convention has to do with Python docstrings. and `%(name)s` falls into this category. 

`%(name)s` is not that commonly used and I don't want people to confuse this for a Mathics3 kind of pattern. 

This cross-level confusion we see in the homegrown docstring doctests as well, where `\` before a newline is a Python string thing, but it is embedded within the homegrown doc system. 

`%(name)s` is more suited when there is a changes the class or may change and we want to adjust this automatically.  However the chance that `TextSentence` will change is not likely as it is a WMA builtin. And if it does, it probably good to go over everything explicitly. 

Again, in short we have had too much cleverness and not a lot of clarity. I think we should try clarity first, and cleverness when it is really warranted which is not here.

File: pymathics/natlang/main.py (hunk 1 above)
Comment (rocky) [reply pymathics/natlang/main.py]: Thanks - fixed now

File: pymathics/natlang/main.py (hunk 2 above)
Comment (mmatera) [reply pymathics/natlang/main.py:373]: Actually, the argument in `OptionsPattern` is not needed here.

File: pymathics/natlang/main.py (hunk 7 above)
Comment (mmatera) [reply pymathics/natlang/main.py:1253]: `head` is an attribute for `Expression`, while `get_head` is a method in all the daughter classes of `BaseElement`

Comment (rocky) [reply pymathics/natlang/main.py:1253]: On Atoms we don't have a "head" property. We should. Maybe  after 6.0.0 we will correct that.

File: pymathics/natlang/main.py (hunk 3 above)
Comment (rocky) [reply pymathics/natlang/main.py:1168]: When not everything has been converted which is what happens here. I don't have the patience to figure how to get converted so that ListExpression can be used. 

When I wrote in the top-level comment: 

> Probably the code be revised so that to_expression, and to_mathics_list are not needed. That is deferred for another PR.

File: pymathics/natlang/main.py (hunk 5 above)
Comment (rocky) [reply pymathics/natlang/main.py]: This and a number of other places have been corrected.

File: pymathics/natlang/main.py (hunk 6 above)
Comment (rocky) [reply pymathics/natlang/main.py]: This and other places have now been corrected.

File: pymathics/natlang/main.py (hunk 2 above)
Comment (rocky) [reply pymathics/natlang/main.py:373]: Ok - after this lands, if you are up for it  would you go in and remove all the the `OptionsPattern`s that are not needed? 

Also if you are up for it remove the `to_expression`s and `to_lists` by adding the additional conversions where needed.
